mantidproject/mantid
# PlotSelector Widget additional features

**Description of work.**

This adds a number of features to the workbench plot selector:

* Right-click menu option to close plots
* Plots can be closed by clicking the X next to a plot name
* Right-click menu option to rename plot
* Option to sort by either name or last accessed plot, ascending or descending
* Export to export plots a useful subset of supported types
* Right-click menu to export plot
* Arrange option to tile or cascade the current open plots (this could be investigated further to see if anything nicer can be done here to arrange the plots)
* Select All option
* Checkable plot names to make multiple selection easier
* Edit button next to close button on plot name, to rename plot

Additionally changes here should make the plot selector more robust to changes in the global figure manager.

**To test:**

This requires some thorough functional testing on Windows, Mac and Linux to check the behaviour, and a code review.

Some things to try:
* Make lots of plots and test performance
* Manipulate plots from matplotlib directly and check behaviour is reflected in plot selector
* Check everything behaves as correctly when multiple features are used, e.g. try setting a filter, sorting by last shown and exporting plots

A useful way to make lots of plots:
```python
for i in range(5):
    plt.figure()
    plt.plot([1 ,2*i, 3*i])
```

**Some open questions about the plot selector and plotting, probably for separate tickets:**
* Should we ensure figure titles are unique, how to do this?
  * **A.** Figure titles not unique, plot number shown as unique identifier
* What plot name/figure title characters are and are not valid? Currently plot names will be used for file names when exporting, should we tidy up then or enforce a limited character set? This needs testing on Windows which is most restrictive on characters.
  * **A.** Anything allowed, special characters replaced when saving
* Performance of saving - this is slow for vector types with large plots
  * See issue https://github.com/mantidproject/mantid/issues/22992
* Plots can be created without being shown, e.g. not calling `figure.show()`. These are shown in the plot selector and can be opened up, but clicking close afterwards destroys the figure completely. Should we show these in the plot selector? Should we have a way to hide them again?
  * **A.** Plots can be hidden
* No warnings are given if overwriting when exporting, is this ok? Same behaviour as for exporting from matplotlib.
  * **A.** Yes, this is ok!

Fixes #22518 
Fixes #22995

Does this update require release notes?


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
